spec/mocks/mock-api.ts

```diff
@@ -0,0 +1,183 @@
+import { vi } from "vitest"
+import { jsonMockData } from "./mock-data"
+
+type ProductTagging = {
+  product_id: string
+  selected_sku_id?: string
+}
+
+type Customer = {
+  email: string
+  first_name: string
+  last_name: string
+  type: string
+}
+
+type Data = {
+  elements?: string[]
+  responseMode: string
+  variation?: string
+  pageType?: string
+  products: ProductTagging[]
+  categories?: string[]
+  cart?: unknown
+  customer?: Customer
+  searchTerms?: string[]
+}
+
+function normaliseProduct(data: ProductTagging | string) {
+  return typeof data === "string" ? { product_id: data } : data
+}
+
+function getPageAction(pageType: string, placements: string[], state: Data) {
```

Comment:
can you try to extract a newAction function?
there is too much structural repetition in this function